internal/resource/opener.go

```diff
@@ -47,153 +79,103 @@ func NewResourceOpenerForApplication(
-	}, "", applicationName)
-}
-
-// noopDownloadResourceLocker is a no-op download resource locker.
-type noopDownloadResourceLocker struct{}
-
-// Acquire grabs the lock for a given application so long as the
-// per application limit is not exceeded and total across all
-// applications does not exceed the global limit.
-func (noopDownloadResourceLocker) Acquire(string) {}
-
-// Release releases the lock for the given application.
-func (noopDownloadResourceLocker) Release(appName string) {}
-
-type resourceClientGetterFunc func(ctx context.Context) (*ResourceRetryClient, error)
-
-func newInternalResourceOpener(
-	args ResourceOpenerArgs,
-	resourceDownloadLimiterFunc func() ResourceDownloadLock,
-	unitName, appName string,
-) (opener resource.Opener, err error) {
-	var unit *state.Unit
-	if unitName != "" {
-		unit, err = args.State.Unit(unitName)
-		if err != nil {
-			return nil, errors.Annotate(err, "loading unit")
-		}
-	}
-
-	if appName == "" {
-		if unit == nil {
-			return nil, errors.Errorf("missing both unit and application")
-		}
-		appName = unit.ApplicationName()
-	}
-	application, err := args.State.Application(appName)
+	application, err := args.State.Application(applicationName)
 	if err != nil {
-		return nil, errors.Trace(err)
+		return nil, errors.Capture(err)
 	}
 
-	var chURLStr *string
-	if unit != nil {
-		chURLStr = unit.CharmURL()
-	} else {
-		chURLStr, _ = application.CharmURL()
-	}
+	chURLStr, _ := application.CharmURL()
 	if chURLStr == nil {
-		return nil, errors.Errorf("missing charm URL for %q", appName)
+		return nil, errors.Errorf("missing charm URL for %q", applicationName)
 	}
+
 	charmURL, err := charm.ParseURL(*chURLStr)
 	if err != nil {
-		return nil, errors.Trace(err)
+		return nil, errors.Capture(err)
 	}
 
+	return &ResourceOpener{
+		state:                args.State.Resources(args.Store),
+		modelUUID:            args.State.ModelUUID(),
+		resourceClientGetter: newClientGetter(charmURL, args.ModelConfigService),
+		retrievedBy:          application.Tag(),
+		charmURL:             charmURL,
+		charmOrigin:          *application.CharmOrigin(),
+		appName:              applicationName,
+		unitName:             "",
+		resourceDownloadLimiterFunc: func() ResourceDownloadLock {
+			return noopDownloadResourceLocker{}
+		},
+	}, nil
+}
+
+func newClientGetter(charmURL *charm.URL, modelConfigService ModelConfigService) resourceClientGetterFunc {
 	var clientGetter resourceClientGetterFunc
 	switch {
 	case charm.CharmHub.Matches(charmURL.Schema):
-		clientGetter = newCharmHubOpener(args.ModelConfigService)
+		clientGetter = newCharmHubOpener(modelConfigService)
 	default:
-		// Use the nop opener that performs no store side requests. Instead it
-		// will resort to using the state package only. Any thing else will call
+		// Use the nop opener that performs no store side requests. Instead, it
+		// will resort to using the state package only. Anything else will call
 		// a not-found error.
```

Comment:
todo: update comment to be in line with actual nopopener.

side note: we have two spellings of no-op in this file, `nop` and `noop`, it'd be nice to consolidate as you're doing some refactor here.

Reply:
True, have changed all to `noop`